pkg/tsdb/cloudwatch/models/cloudwatch_query.go

```diff
@@ -390,21 +400,27 @@ func getLabel(query metricsDataQuery, dynamicLabelsEnabled bool) string {
 	return result
 }
 
+func calculatePeriodBasedOnTimeRange(startTime, endTime time.Time) int {
+	deltaInSeconds := endTime.Sub(startTime).Seconds()
+	periods := getRetainedPeriods(time.Since(startTime))
+	datapoints := int(math.Ceil(deltaInSeconds / 2000))
```

Comment:
what does 2000 represent here? maybe it can be a named variable or done in some kind of math to indicate what's happening (example 7 * 24 for hours in a week)

Reply:
I would if I could, but I have no idea what it represents. Would guess it's just a random quotient that seemed suitable at the time this was implemented..?

Reply:
So this number comes from CloudWatch?
I would at least add a link to a reference in this case.

Reply:
No I did not say it comes from CloudWatch.

Reply:
Ah sorry... I misunderstood. I thought "seemed suitable at the time it was implemented" referred to a decision made by an AWS engineer during CloudWatch development.
So, where does the logic behind this line of code come from?

Reply:
Don't know actually.

Reply:
Let's try to trace it, then.
1. https://github.com/grafana/grafana/pull/57624/files#diff-870ca91de9ab6432ae4482009df9ce24421b3a4aeb334e94a61bb13285a865baR384
2. https://github.com/grafana/grafana/pull/21659/files#diff-85dc934cbe9964e970a0c2745dffcf54bec2f1e61d505534fcb7f437207d900dR72
3. https://github.com/grafana/grafana/commit/d40b66f1c170d0927382854f9b36351937c05757#diff-85dc934cbe9964e970a0c2745dffcf54bec2f1e61d505534fcb7f437207d900dR72
So, this was introduced by https://github.com/grafana/grafana/pull/21471 .

Reply:
The documentation says:
> * Data points with a period of less than 60 seconds are available for 3 hours. These data points are high-resolution metrics and are available only for custom metrics that have been defined with a StorageResolution of 1.
> * Data points with a period of 60 seconds (1-minute) are available for 15 days.
> * Data points with a period of 300 seconds (5-minute) are available for 63 days.
> * Data points with a period of 3600 seconds (1 hour) are available for 455 days (15 months).

The original code was (abbreviated):
```go
func closest(array []int, num int) int {
	if num <= array[0] -> return array[0]
	if num >= array[len(array)-1]  -> return array[len(array)-1]
	minDiff := array[len(array)-1]
	var closest int
	for _, value := range array {
		var m = int(math.Abs(float64(num - value)))
		if m <= minDiff  -> minDiff = m; closest = value
	}
	return closest
}
periods := []int{60, 300, 900, 3600, 21600}
period = closest(periods, int(math.Ceil(deltaInSeconds/2000)))
```

I deduced that "closeness" to two consecutive numbers in the periods table (i.e. `min|dTs/2000-p[i]|`) means a break at each midpoint. So, this would result in the following numbers:
```
dTs <= 2000*(60+300)/2=36,000=10h -> p=60
10h < dTs <= 2000*(300+900)/2=1,200,000=13.89d -> p=300
13.89d < dTs <= 2000*(900+3600)/2=52.08d -> p=900
52.08d < dTs <= 2000*(3600+21600)/2=25,200,000=291.67d -> p=3600
291.67d < dTs <= 2000*21600=43,200,000=500d -> p=21600
```

That still doesn't explain where this magic number comes from, though. I can only guess that it is some sort of hard-coded number of data points.

But the only "magic" number that is really relevant, is 10,800: That's the maximum number of data points a GetMetricData query will return. The only relation to this number I can find is `2000*21,600=10,800,000=10,800/2`, but that could be a red herring.

Reply:
Thanks for the thorough investigation! 

Interesting that this was initially introduced in a PR made by me - I have no memory of this whatsoever. My guess is that the number `2000` is used in the CloudWatch console, so we were recommended to use the same in Grafana. But I'm not sure. 

There's no point in asking the GetMetricData api for more data points than the graph has pixels for. My guess is that the number `2000` is an assumption made for available pixels in the graph. However, the query data api already provides a `MaxDataPoints` field for each query, so we should probably use that instead of `2000`. That would be a breaking change though as it may change result set for potentially many user, so that's something that we need to save for a major release in that case.

Reply:
> what does 2000 represent here?

![image](https://user-images.githubusercontent.com/333884/208692370-85e59164-dc70-44c4-8180-b992c28d9858.png)

https://docs.aws.amazon.com/grafana/latest/userguide/CloudWatch-using-the-metric-search.html#cloudwatch-period

My 2 cent...

Reply:
@piequi that paragraph is copied from the cloudwatch data source documentation at grafana.com. 

The question here was why ‘2000’ was added to the equation.

Reply:
Sorry for bringing this up and making your past catch up with you. :smile: 
I suppose it's ok to leave this as it is for now, as it's unrelated to the PR. The current algorithm looks better than the original version, in that it no longer breaks at the midpoints instead of the boundaries (given that my interpretation was correct).
But it may not be a bad idea to revisit this at some point.